app/controllers/application_controller/compare.rb

```diff
@@ -748,44 +748,47 @@ def compare_init(mode)
     @exists_mode = session[:miq_exists_mode]
   end
 
-  # Compare selected VMs
+  # Compare selected items
   def comparemiq
     assert_privileges(params[:pressed])
-    vms = []
+    items = []
     if !session[:checked_items].nil? && @lastaction == "set_checked_items"
-      vms = session[:checked_items]
+      items = session[:checked_items]
     else
-      vms = find_checked_items
+      items = find_checked_items
     end
 
-    case request.parameters["controller"].downcase
-    when "ems_cluster"
-      title = _("Clusters")
-    when "vm"
-      title = _("Virtual Machines")
-    when "miq_template"
-      title = _("VM Templates")
-    else
-      title = request.parameters["controller"].pluralize.titleize
-    end
-    if vms.length < 2
+    title = case request.parameters["controller"].downcase
+            when "ems_cluster"
+              _("Clusters")
+            when "vm"
+              _("Virtual Machines")
+            when "miq_template"
+              _("VM Templates")
+            else
+              request.parameters["controller"].pluralize.titleize
+            end
+
+    if items.length < 2
       add_flash(_("At least 2 %{model} must be selected for Compare") % {:model => title}, :error)
       if @layout == "vm" # In vm controller, refresh show_list, else let the other controller handle it
         show_list
         @refresh_partial = "layouts/gtl"
       end
-    elsif vms.length > 32
+    elsif items.length > 32
       add_flash(_("No more than 32 %{model} can be selected for Compare") % {:model => title}, :error)
       if @layout == "vm" # In vm controller, refresh show_list, else let the other controller handle it
         show_list
         @refresh_partial = "layouts/gtl"
       end
     else
-      session[:miq_selected] = vms        # save the selected vms array for the redirect to compare_miq
       if params[:pressed]
         model, = pressed2model_action(params[:pressed])
-        @sb[:compare_db] = compare_db(model)
+        klass_name = compare_db(model)
+        @sb[:compare_db] = klass_name
+        assert_rbac(klass_name.constantize, items)
       end
+      session[:miq_selected] = items        # save the selected items array for the redirect to compare_miq
```

Comment:
we have `session[:checked_items]` and session[:miq_selected]`

this code is giving me a headache